sound/soc/soc-pcm.c

```diff
@@ -2918,30 +2919,41 @@ int soc_new_pcm(struct snd_soc_pcm_runtime *rtd, int num)
 	}
 
 	if (rtd->dai_link->dynamic || rtd->dai_link->no_pcm) {
-		for_each_rtd_cpu_dais(rtd, i, cpu_dai) {
+		if (rtd->dai_link->dpcm_playback) {
+			invalid_dais = 0;
+			stream = SNDRV_PCM_STREAM_PLAYBACK;
+
+			for_each_rtd_cpu_dais(rtd, i, cpu_dai)
+				if (!snd_soc_dai_stream_valid(cpu_dai, stream))
+					invalid_dais++;
 
-			if (rtd->dai_link->dpcm_playback) {
-				stream = SNDRV_PCM_STREAM_PLAYBACK;
-				if (snd_soc_dai_stream_valid(cpu_dai, stream)) {
-					playback = 1;
-				} else if (playback != 0) {
-					dev_err(rtd->card->dev,
-						"CPU DAI %s does not support playback\n",
-						cpu_dai->name);
-					return -EINVAL;
-				}
+			if (!invalid_dais) {
+				playback = 1;
+			} else {
+				dev_err(rtd->card->dev,
+					"%d invalid CPU DAI(s) for rtd %s without playback support\n",
+					invalid_dais,
+					rtd->dai_link->stream_name);
+				return -EINVAL;
```

Comment:
There are another two flags: playback_only and capture_only (L2986). How do they work with dpcm_playback and dpcm_capture ? The original control flow will go through to L2986, but now it changes. I am really confused that whether dpcm_playback and dpcm_capture can coexist or not. The original code checks both flags.
````	
if (rtd->dai_link->playback_only) {
		playback = 1;
		capture = 0;
	}

if (rtd->dai_link->capture_only) {
	playback = 0;
	capture = 1;
}
````

Reply:
Another question:  for max98373 DSM case we need both playback and capture dai in one pcm?

Reply:
> There are another two flags: playback_only and capture_only (L2986). How do they work with dpcm_playback and dpcm_capture ? The original control flow will go through to L2986, but now it changes. I am really confused that whether dpcm_playback and dpcm_capture can coexist or not. The original code checks both flags.
> 
> ```
> if (rtd->dai_link->playback_only) {
> 		playback = 1;
> 		capture = 0;
> 	}
> 
> if (rtd->dai_link->capture_only) {
> 	playback = 0;
> 	capture = 1;
> }
> ```

ok, I had absolutely no idea those flags even existed, will look into this. Thanks for the feedback @RanderWang

Reply:
> Another question: for max98373 DSM case we need both playback and capture dai in one pcm?

the max98373 provides feedback but the capture is split between echo reference and IV feedback, so the PCM (on the host side) is likely different.

Reply:
@RanderWang I asked the question on alsa-devel and no one knows what those flags are, so I'd like to go ahead with this fix and revisit why we have different flags later. This PR does not add new issues that weren't there already. Thanks!